### PR TITLE
Update masking for cudann attention

### PR DESCRIPTION
**Pull request by absternator**
## Summary
This PR makes an update to the masking strategy for the cudanns flash attention. This is to avoid OOMs for large number of sites.

### New Features
- A dense mask is no longer created. Instead a 1D array of (batch,) is passed into query_seq_lengths, key_seq_lengths

### Fixes

---

### All PRs
- [ ] **Correctness**: Logic is sound and implementation is correct
- [ ] **Testing**: Appropriate tests added/updated
- [ ] **Benchmarking**: Training/sampling speed impact assessed if relevant
- [ ] **Documentation**: Code comments and docs updated where needed

### For AI-Generated PRs
- [ ] **Entropy**: No unnecessary code duplication, verbose implementations,
      or misplaced imports
- [ ] **Test Quality**: Tests genuinely validate intended behavior (check
      `.claude/spec/{feature}/tasks.md` for requirements)
- [ ] **Steering Docs**: Fundamental changes to the repo (if any) are
        reflected in steering documents `.claude/steering/`

### CODE REVIEW

File: tfmpe/nn/transformer/transformer.py [hunk 1]
```diff
@@ -126,11 +126,13 @@ def encode(
...
         # Apply encoder blocks sequentially via scan
         if tokens.padding_mask is not None:
-            sample_shape = tokens.sample_shape
-            n_tokens = tokens.data.shape[tokens.sample_ndims]
-            n_heads = self.config.n_heads
-            mask = tokens.padding_mask[:, None, None, :]
-            mask = jnp.broadcast_to(mask, sample_shape + (n_heads, n_tokens, n_tokens))
+            if self.config.attention == 'cudnn':
+                # Pass (batch,) seq lengths — varlen flash attention, no dense mask
+                mask = jnp.sum(tokens.padding_mask, axis=-1).astype(jnp.int32)
```
Comment (giovannic) [inline tfmpe/nn/transformer/transformer.py]: Nice! Could we do this once in Tokens instead of every forward pass? We can have something like Tokens.seq_len which is initialised on creation.

**Review by giovannic — CHANGES_REQUESTED** [review]
Thanks for finding this! Just pointing to a neater/more efficient place to put the change

File: tfmpe/nn/transformer/transformer.py (hunk 1 above)
Comment (absternator) [reply tfmpe/nn/transformer/transformer.py]: added total_tokens field that can be used as suggested

EDIT: added func attention_mask instead as cant hard code the total tokens as when combine_tokens occurs this can mess things up

**Review by giovannic — APPROVED** [review]
Fantastic
